C2DEVEL-14695: return backup service support (#54)
C2Devel/terraform-provider-croccloud: website/docs/r/backup_plan.html.markdown

---
subcategory: "Backup"
layout: "aws"
page_title: "aws_backup_plan"
description: |-
  Manages a backup plan.
---

[backup-plan]: https://docs.cloud.croc.ru/en/services/backup/operations.html#backupplan

# Resource: aws_backup_plan

Manages a backup plan. For details about backup plans, see the [user documentation][backup-plan].

## Example Usage

```terraform
resource "aws_backup_vault_default" "example" {}

resource "aws_backup_plan" "example" {
  name = "tf-backup-plan"

  rule {
    rule_name         = "tf-backup-rule"
    target_vault_name = aws_backup_vault_default.example.name
    schedule          = "cron(0 12 * * *)"
    start_window      = 60
    completion_window = 180

    lifecycle {
      delete_after = 30
    }
  }

  rule {
    rule_name         = "tf-backup-rule2"
    target_vault_name = aws_backup_vault_default.example.name
    schedule          = "cron(0 23 * * *)"
    start_window      = 60
    completion_window = 180

    lifecycle {
      delete_after = 60
    }
  }
}
```

## Argument Reference

The following arguments are supported:

* `name` - (Required, Editable) The name of the backup plan.

> * `rule` - (Required, Editable) List of rule objects for separate periodic tasks. The structure of this block is [described below](#rule)

Reviewer: может просто rules ? в следующем предложении пишем, что это блок

### Rule

~> All the parameters in the `rule` block are editable.

The `rule` block has the following structure:

> * `completion_window` - (Optional) Time in minutes after the backup job is started, during which it should be completed. Otherwise, it will be canceled. Defaults to `180`.

Reviewer: cancelled

Author: Поправил

> * `completion_window` - (Optional) Time in minutes after the start of the backup job, during which it should be completed. Otherwise, it will be canceled. Defaults to `180`.

Reviewer: по красоте after the backup job is started, но необ.

> * `lifecycle` - (Optional) The lifecycle defines when a protected resource is transitioned to cold storage and when it expires. The structure of the block is [described below](#lifecycle).

Reviewer: transitioned скорее трансформация
tranferred

> * `rule_name` - (Required) The name of the backup rule. The value must be `1` to `50` characters long and must contain only alphanumeric characters, hyphens, underscores, and periods.

Reviewer: and -> or

* `schedule` - (Optional) CRON expression in UTC for the backup job scheduling.
* `start_window` - (Optional) Time in minutes during which a backup job should start running. Otherwise the job will be cancelled. Defaults to `60`.
* `target_vault_name` - (Required) The name of the backup vault. The value must be `2` to `50` characters long and must contain only alphanumeric characters, hyphens, and underscores.

### Lifecycle

The `lifecycle` block has the following structure:

> * `delete_after` - (Optional, Editable) Number of days from the recovery point creation date, after which it will be deleted.

Reviewer: the date of creation of the restore point

## Attribute Reference

In addition to all arguments above, the following attributes are exported:

* `arn` - The Amazon Resource Name (ARN) of the backup plan.

> * `id` - The backup plan ID.

Reviewer: см. выше

->  **Unsupported attributes**
These attributes are currently unsupported:

> * `advanced_backup_setting` - An object that specifies backup options for each resource type. Always empty.

Reviewer: Specifies ... (начало убрать)

> * `rule` - A rule object that specifies a separate periodic task.

Reviewer: Specifies ...

    * `copy_action` - Configuration block(s) with copy operation settings.

>     * `enable_continuous_backup` - Enable continuous backups for supported resources. Defaults to `false`.

Reviewer: Enables

>     * `lifecycle` - Defines when a protected resource is transferred to cold storage and when it expires.

Reviewer: а тут все правильно? Как сам экземпляр (защищаемый ресурс) может быть перемещен в cold storage?

Author: Т.к. этот атрибут неподдерживаемый, за основу взята формулировка от aws. По идее такие формулировки не особо критичны, пока у нас нет нашей реализации, под которую можно четко описать эти атрибуты. КМК можно оставить и protected resource, если воспринимать это больше как абстракцию, что мы бэкапим этот ресурс и у него есть точки восстановления. И в зависимости от даты последних точек восстановления можно судить насколько протух/актуален бэкапируемый ресурс в нашем хранилище

Author: Обсудили, лучше исправить на recovery point

>     * `lifecycle` - The lifecycle defines when a protected resource is transitioned to cold storage and when it expires.

Reviewer: Defines ...
transferred

>         * `cold_storage_after` - Specifies the number of days after creation that a recovery point is moved to cold storage.

Reviewer: after creation of the recovery point after which it

>     * `recovery_point_tags` - Metadata that you can assign to help organize the resources that you create.

Reviewer: не оч. понятно, что значит organize в этом контексте

Author: У aws по тегам много где такая формулировка фигурирует. Как я понял, они подразумевают под organize управление в плане поиска, фильтрации, группировки ресурсов по их тегам

> * `version` - Unique, randomly generated, Unicode, UTF-8 encoded string that serves as the version ID of the backup plan. Always `""`.

Reviewer: that servers -> used

## Import

> Backup Plan can be imported using `id`, e.g.,

Reviewer: plan

---
subcategory: "Backup"
layout: "aws"
page_title: "aws_backup_plan"
description: |-
  Manages a backup plan resource.
---

> [Backup]: https://docs.cloud.croc.ru/en/services/backup/operations.html#backupplan

Reviewer 2: ```suggestion
[backup-plan]: https://docs.cloud.croc.ru/en/services/backup/operations.html#backupplan
```

На случай, если еще ссылки из этого раздела появятся ) + обновить в месте использования

Author: Поправил

# Resource: aws_backup_plan

> Manages a backup plan resource. For details about backup plans, see the [user documentation][Backup].

Reviewer 2: ```suggestion
Manages a backup plan. For details about backup plans, see the [user documentation][Backup].
```

Author: Поправил

## Example Usage

```terraform
resource "aws_backup_vault_default" "example" {
}

resource "aws_backup_plan" "example" {
  name = "tf-backup-plan"

  rule {
    rule_name         = "tf-backup-rule"
    target_vault_name = aws_backup_vault_default.example.name
    schedule          = "cron(0 12 * * *)"
    start_window      = 60
    completion_window = 180

    lifecycle {
      delete_after = 30
    }
  }
}
```

## Argument Reference

The following arguments are supported:

> * `name` - (Required) The name of the backup plan.

Reviewer 2: Тожe Editable

Author: Поправил

> * `rule` - (Required, Editable) A rule object that specifies a separate periodic task.

Reviewer 2: Тут на самом деле список объектов

В конфигурации будет указываться так:
```
resource "aws_backup_plan" "example" {
  name = "tf-backup-plan"
  
  rule {...}
  
  rule {...}
}
```

Reviewer 2: Для перехода к описанию внутренней структуры блока добавь в конец предложение: `The structure of the block is [described below](#rule).`. 

`#rule` - это ссылка на соответствующий заголовок внизу, там будет `### Rule` вместо `### Rule Arguments`

Author: Добавил еще один блок `rule` в примерах конфигураций + в описании поля `rule`

### Rule Arguments

> For **rule** the following arguments are supported:

Reviewer 2: Используй тут такую формулировку: ``The `rule` block has the following structure:``

Еще до этого предложения можно добавить такой блок, если все параметры редактируемые:
``~> All the parameters in the `rule` block are editable.``

Author: Поправил

* `completion_window` - (Optional) Time in minutes after the start of the backup job, during which it should be completed. Otherwise, it will be canceled. Defaults to `180`.

> * `lifecycle` - (Optional) The lifecycle defines when a protected resource is transitioned to cold storage and when it expires. Fields documented below.

Reviewer 2: ```suggestion
* `lifecycle` - (Optional) The lifecycle defines when a protected resource is transitioned to cold storage and when it expires. The structure of the block is [described below](#lifecycle).
```

Правки как для `rule`

Author: Поправил

> * `rule_name` - (Required) The name of the backup rule. Must be at least `1` character and at most `50` characters in length. Must contain only alphanumeric characters, hyphens, underscores, and periods.

Reviewer 2: ```suggestion
* `rule_name` - (Required) The name of the backup rule. The value must be 1 to 50 characters long and must contain only alphanumeric characters, hyphens, underscores, and periods.
```

Так чуть короче

Author: Поправил

* `schedule` - (Optional) CRON expression in UTC for the backup job scheduling.
* `start_window` - (Optional) Time in minutes during which a backup job should start running. Otherwise the job will be cancelled. Defaults to `60`.

> * `target_vault_name` - (Required) The name of the backup vault. Must be at least `2` characters and at most `50` characters in length. Must contain only alphanumeric characters, hyphens, and underscores.

Reviewer 2: ```suggestion
* `target_vault_name` - (Required) The name of the backup vault. The value must be 2 to 50 characters long and must contain only alphanumeric characters, hyphens, and underscores.
```

Author: Поправил

> ### Lifecycle Arguments

Reviewer 2: см. комментарии для rule

Author: Поправил

## Attribute Reference

In addition to all arguments above, the following attributes are exported:

* `arn` - The Amazon Resource Name (ARN) of the backup plan.
* `id` - The backup plan ID.

->  **Unsupported attributes**
These attributes are currently unsupported:

* `advanced_backup_setting` - An object that specifies backup options for each resource type. Always empty.
* `rule` - A rule object that specifies a separate periodic task.
    * `copy_action` - Configuration block(s) with copy operation settings.
    * `enable_continuous_backup` - Enable continuous backups for supported resources. Defaults to `false`.
    * `lifecycle` - The lifecycle defines when a protected resource is transitioned to cold storage and when it expires. Fields documented below.
        * `cold_storage_after` - Specifies the number of days after creation that a recovery point is moved to cold storage.
    * `recovery_point_tags` - Metadata that you can assign to help organize the resources that you create.
* `version` - Unique, randomly generated, Unicode, UTF-8 encoded string that serves as the version ID of the backup plan. Always `""`.

## Import

> Backup Plan can be imported using the `id`, e.g.,

Reviewer 2: ```suggestion
Backup Plan can be imported using `id`, e.g.,
```

Author: Поправил

---
subcategory: "Backup"
layout: "aws"
page_title: "aws_backup_plan"
description: |-
  Manages an backup plan resource.
---

[Backup]: https://docs.cloud.croc.ru/en/services/backup/operations.html#backupplan

# Resource: aws_backup_plan

Manages an backup plan resource. For details about backup plans, see the [user documentation][Backup].

## Example Usage

```terraform
resource "aws_backup_vault" "example" {
}

resource "aws_backup_plan" "example" {
  name = "tf-backup-plan"

  rule {
    rule_name         = "tf-backup-rule"
    target_vault_name = aws_backup_vault.example.name
    schedule          = "cron(0 12 * * *)"
    start_window      = 60
    completion_window = 180

    lifecycle {
      delete_after = 30
    }
  }
}
```

## Argument Reference

The following arguments are supported:

* `name` - (Required) The name of the backup plan.
* `rule` - (Required, Editable) A rule object that specifies a separate periodic task.

> ->  **Unsupported arguments**

Reviewer 2: Я их обычно в раздел `Unsupported attributes` перемещала, тк формально они тут описываются только потому что они есть в итоговом состоянии. И получался один такой раздел на одну страничку в документации

Author: Поправил

    rule_name         = "tf_example_backup_rule"
    target_vault_name = aws_backup_vault.test.name

>     schedule          = "cron(0 12 * * ? *)"

Reviewer 2: Падает с ошибкой `InvalidParameterValue: Cron expression is not valid: There should be exactly 5 fields.`

Author: Поправил экзамплы, проверил локально

>     start_window             = 60

Reviewer 2: Прогони таргет ` make website-lint-fix`, он поправит ошибки линтинга (тут выровняет конфигурацию)

Author: поправил

* `name` - (Required) The display name of a backup plan.
* `rule` - (Required) A rule object that specifies a scheduled task that is used to back up a selection of resources.

> * `advanced_backup_setting` - (Optional) An object that specifies backup options for each resource type. _(Currently unsupported)_.

Reviewer 2: Неподдерживаемые параметры оформляются иначе. Как пример оформления доки можно использовать `ec2_transit_gateway.html.markdown`

Формулировки для описания параметров можно брать из нашей документации, обычно они более четко сфомулированы.

Author: Поправил

* `tags` - (Optional) Metadata that you can assign to help organize the plans you create. If configured with a provider [`default_tags` configuration block](/docs/providers/aws/index.html#default_tags-configuration-block) present, tags with matching keys will overwrite those defined at the provider-level. _(Currently unsupported)_.

### Rule Arguments
For **rule** the following attributes are supported:

> * `rule_name` - (Required) An display name for a backup rule.

Reviewer 2: У поля есть валидации в схеме на длину и регехп. Они актуальны для бэкап апи (в доке инфы нет)?
Если актуальны, надо добавить инфу о них в документацию.

Аналогично для  `target_vault_name`. Там они вообще кажется не нужны, тк это имя существующего волта.

Author: Добавил инфу по валидациям

* `target_vault_name` - (Required) The name of a logical container where backups are stored.
* `schedule` - (Optional) A CRON expression specifying when AWS Backup initiates a backup job.
* `enable_continuous_backup` - (Optional) Enable continuous backups for supported resources. _(Currently unsupported)_.

> * `start_window` - (Optional) The amount of time in minutes before beginning a backup.

Reviewer 2: Для поддерживаемых **аргументов**, которые имеют дефолтные значения в схеме, в конце описания добаляется `Defaults to <value>.`

Author: Добавил

* `delete_after` - (Optional) Specifies the number of days after creation that a recovery point is deleted. Must be 90 days greater than `cold_storage_after`.

### Copy Action Arguments
For **copy_action** the following attributes are supported:

* `lifecycle` - (Optional) The lifecycle defines when a protected resource is copied over to a backup vault and when it expires.  Fields documented above. _(Currently unsupported)_.
* `destination_vault_arn` - (Required) An Amazon Resource Name (ARN) that uniquely identifies the destination backup vault for the copied backup. _(Currently unsupported)_.

### Advanced Backup Setting Arguments
For `advanced_backup_setting` the following attibutes are supported:

* `backup_options` - (Required) Specifies the backup option for a selected resource. This option is only available for Windows VSS backup jobs. Set to `{ WindowsVSS = "enabled" }` to enable Windows VSS backup option and create a VSS Windows backup. _(Currently unsupported)_.
* `resource_type` - (Required) The type of AWS resource to be backed up. For VSS Windows backups, the only supported resource type is Amazon EC2. Valid values: `EC2`. _(Currently unsupported)_.

> ## Attributes Reference

Reviewer 2: ```suggestion
## Attribute Reference
```

Так для каждой странички

Author: Поправил
